File: O/oexample.cpp
// principio Open/Closed

#include <iostream>
using namespace std;

// Clase base
class Pinguino {
public:
    virtual void actuar() = 0;  // Método que puede extenderse
    virtual ~Pinguino() {}
};

// Pingüino que actúa en aguas superficiales
class PinguinoSuperficial : public Pinguino {
public:
    void actuar() override {
        cout << "Actúo en aguas superficiales." << endl;
    }
};

// Pingüino que actúa en aguas profundas
class PinguinoProfundo : public Pinguino {
public:
    void actuar() override {
        cout << "Actúo en aguas profundas." << endl;
    }
};

// Función que usa cualquier tipo de pingüino sin modificarla
void mostrarAccion(Pinguino* p) {
    p->actuar();
}

int main() {
    PinguinoSuperficial superficial;
    PinguinoProfundo profundo;

    mostrarAccion(&superficial); // ✅ "Actúo en aguas superficiales."
    mostrarAccion(&profundo);    // ✅ "Actúo en aguas profundas."

    return 0;
}

File: O/oexample.py
# principio Open/Closed

from abc import ABC, abstractmethod

# Clase base abstracta
class Pinguino(ABC):
    @abstractmethod
    def actuar(self):
        pass

# Pingüino que actúa en aguas superficiales
class PinguinoSuperficial(Pinguino):
    def actuar(self):
        print("Actúo en aguas superficiales.")

# Pingüino que actúa en aguas profundas
class PinguinoProfundo(Pinguino):
    def actuar(self):
        print("Actúo en aguas profundas.")

# Nueva clase sin modificar nada anterior
class PinguinoMixto(Pinguino):
    def actuar(self):
        print("Actúo en aguas superficiales.")
        print("Actúo en aguas profundas.")

# Función que puede trabajar con cualquier tipo de pingüino
def mostrar_accion(pinguino: Pinguino):
    pinguino.actuar()

# Ejemplo de uso
ping1 = PinguinoSuperficial()
ping2 = PinguinoProfundo()
ping3 = PinguinoMixto()

mostrar_accion(ping1)  # ✅ "Actúo en aguas superficiales."
mostrar_accion(ping2)  # ✅ "Actúo en aguas profundas."
mostrar_accion(ping3)  # ✅ "Actúo en aguas superficiales." y "Actúo en aguas profundas."
